### PR TITLE
HIP-584: Implement `DissociatePrecompile`

### LINKED ISSUE #5648: HIP-584: Provide implementation for associate and dissociate precompiles
### Problem

We need default implementations for precompiles related to associate and dissociate logic, so we should copy needed logic from hedera-services into hedera-mirror-web3.

### Solution

Copy precompile logic for the following precompiles:

- AbstractAssociatePrecompile
- AssociatePrecompile
- AbstractDissociatePrecompile
- DissociatePrecompile

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by stoyanov-st**
**Description**:

Implementing `DissociatePrecompile` using `DisssociateLogic`.

**Related issue(s)**:

Fixes #5648

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/DissociatePrecompile.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    @Override
+    public long getGasRequirement(long blockTimestamp, Builder transactionBody, Store store) {
+        return pricingUtils.computeGasRequirement(blockTimestamp, this, transactionBody, store);
+    }
+
+    @Override
+    public Set<Integer> getFunctionSelectors() {
+        return Set.of(ABI_ID_DISSOCIATE_TOKEN);
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/DissociatePrecompile.java:82]: We should add `ABI_ID_HRC_DISSOCIATE` and `ABI_ID_DISSOCIATE_TOKENS`

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServicePrecompileTest.java [hunk 2]
```diff
@@ -359,7 +357,11 @@ enum UnsupportedContractModificationFunctions {
     enum SupportedContractModificationFunctions {
         ASSOCIATE_TOKEN("associateTokenExternal", new Object[] {SPENDER_ADDRESS, FUNGIBLE_TOKEN_ADDRESS}),
         ASSOCIATE_TOKENS(
-                "associateTokensExternal", new Object[] {SPENDER_ADDRESS, new Address[] {FUNGIBLE_TOKEN_ADDRESS}});
+                "associateTokensExternal", new Object[] {SPENDER_ADDRESS, new Address[] {FUNGIBLE_TOKEN_ADDRESS}}),
+        DISSOCIATE_TOKEN("dissociateTokenExternal", new Object[] {SPENDER_ADDRESS, NOT_FROZEN_FUNGIBLE_TOKEN_ADDRESS}),
+        DISSOCIATE_TOKENS(
+                "dissociateTokensExternal",
+                new Object[] {SPENDER_ADDRESS, new Address[] {NOT_FROZEN_FUNGIBLE_TOKEN_ADDRESS}});
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServicePrecompileTest.java:364]: We can add test cases for `HRC_ASSOCIATE` and `HRC_DISSOCIATE` by adding new methods in the smart contract used for testing.

File: hedera-mirror-web3/src/main/java/com/hedera/services/fees/calculation/token/txns/TokenDissociateResourceUsage.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import com.hederahashgraph.api.proto.java.TransactionBody;
+import java.util.function.BiFunction;
+import javax.inject.Inject;
+
+public class TokenDissociateResourceUsage extends AbstractTokenResourceUsage implements TxnResourceUsageEstimator {
+    private static final BiFunction<TransactionBody, TxnUsageEstimator, TokenDissociateUsage> factory =
+            TokenDissociateUsage::newEstimate;
+
+    @Inject
+    public TokenDissociateResourceUsage(final EstimatorFactory estimatorFactory) {
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/fees/calculation/token/txns/TokenDissociateResourceUsage.java:40]: You can remove @Inject annotation

**Review by IvanKavaldzhiev — DISMISSED** [review]
LGTM! The only thing missing is a bit more test cases in `DissociatePrecompileTest`

File: hedera-mirror-web3/src/main/java/com/hedera/services/fees/calculation/token/txns/TokenDissociateResourceUsage.java (hunk 3 above)
Comment (stoyanov-st) [reply hedera-mirror-web3/src/main/java/com/hedera/services/fees/calculation/token/txns/TokenDissociateResourceUsage.java:38]: Noted

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/DissociatePrecompile.java (hunk 1 above)
Comment (stoyanov-st) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/DissociatePrecompile.java]: We will introduce it in a separate PR.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServicePrecompileTest.java (hunk 2 above)
Comment (stoyanov-st) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServicePrecompileTest.java]: We will introduce the tests in a separate PR.

File: hedera-mirror-web3/src/test/java/com/hedera/services/store/contracts/precompile/impl/DissociatePrecompileTest.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        given(hrcParams.token()).willReturn(tokenID);
+        given(hrcParams.senderAddress()).willReturn(callerAccountAddress);
+
+        final var accountID =
+                EntityIdUtils.accountIdFromEvmAddress(Objects.requireNonNull(callerAccountAddress.toArray()));
+        final var expected = syntheticTxnFactory.createDissociate(Dissociation.singleDissociation(accountID, tokenID));
+        final var result = dissociatePrecompile.body(dissociateToken, a -> a, hrcParams);
+
+        assertEquals(expected.getTokenDissociate(), result.getTokenDissociate());
+    }
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/test/java/com/hedera/services/store/contracts/precompile/impl/DissociatePrecompileTest.java:79]: You can copy and modify the unit tests for `DissociatePrecompile` from `hedera-services` like we did for `AssociatePrecompile`.

**Review by steven-sheehy — APPROVED** [review]

**Pull request merged**